sdk/servicebus/service-bus/src/receivers/bufferReceiver.ts

```diff
@@ -0,0 +1,107 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+import { Receiver, ReceiverEvents, EventContext } from "rhea-promise";
+import { DefaultDataTransformer } from "@azure/core-amqp";
+
+export function messageIterator(
+  receiver: Receiver,
+  initialCredits: number
+): AsyncIterableIterator<string> & { close(): Promise<void> } {
+  const iterator = _iterator(receiver, initialCredits) as ReturnType<typeof messageIterator>;
+
+  iterator.close = async () => {
+    return new Promise((resolve) => {
+      receiver.drain = true;
+      receiver.addCredit(1);
+
+      receiver.once(ReceiverEvents.receiverDrained, () => {
+        resolve();
+      });
+    });
```

Comment:
I do like this. What happens if we get any events while draining? I'm not sure how I'd get access to those (is it ridiculous to return any events received at this point? We're not closing the link so they could still be settled right?)

Reply:
It's not ridiculous - right now the semantics around drain are more unclear than I'd like. But I think what  you want to do make sense - just return what we have left. 

The only lame thing is that I've constrained the promise to only return a single event at a time. That could get fixed.

Reply:
Another thing that I've been thinking about is whether or thinking about 'drain' being the "final" event is correct or not. So we need to get confirmation on that as well (if so, returning the last few events here would be even more amazing)